### ISSUE TITLE
Document what a "dangling pointer" is

### ISSUE THREAD

**Issue by Urgau** [rust-lang/rust#132286]
## Location

There are several places this applies, e.g.:

 - [`std::ptr::dangling`](https://doc.rust-lang.org/nightly/std/ptr/fn.dangling.html)
 - [`std::ptr::from_ref`](https://doc.rust-lang.org/nightly/std/ptr/fn.from_ref.html) *(in the examples)*
 - `dangling_pointers_from_temporaries` lint

## Summary

We have several place in the standard library or as part of lints where we use the term "dangling pointer" without defining it anywhere.

We should have a place where we define what that is, ideally in the standard library, in it's own section, maybe under `std::ptr`, so it can be referenced elsewhere.

It should probably have a suggest definition, followed by some bad examples (and maybe how to fix them).

[Wikipedia has article](https://en.wikipedia.org/wiki/Dangling_pointer), *I don't know if it could serve as a base our documentation but still worth checking it.*

## Steps

 - [ ] Add std documentation
 - [ ] Update existing std docs
 - [ ] Update lints diagnostics

**Comment by saethlin** [comment]
cc @rust-lang/opsem

**Comment by RalfJung** [comment]
Also see https://github.com/rust-lang/unsafe-code-guidelines/issues/478

**Comment by theemathas** [comment]
[The reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers) has a definition of dangling pointers.

**Comment by RalfJung** [comment]
Right, so somewhat confusingly `ptr::dangling` on a ZST does actually produce a non-dangling pointer.

**Comment by theemathas** [comment]
The documentation for `ptr::dangling` even explicitly says that it produces a dangling pointer (which is false). And then it goes to say that "the pointer value may potentially represent a valid pointer to a `T`" (which is false for non-ZST types, due to provenance).

**Comment by theemathas** [comment]
Documentation from [the reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#dangling-pointers):
> A reference/pointer is “dangling” if not all of the bytes it [points to](https://doc.rust-lang.org/reference/behavior-considered-undefined.html#pointed-to-bytes) are part of the same live allocation (so in particular they all have to be part of some allocation).
> If the size is 0, then the pointer is trivially never “dangling” (even if it is a null pointer).

This implies that a pointer points to a range of addresses, depending on the type, and whether a pointer is dangling depends on the type.

Documentation from [`std::ptr`](https://doc.rust-lang.org/nightly/std/ptr/index.html#safety):
> We say that a pointer is “dangling” if it is not valid for any non-zero-sized accesses. This means out-of-bounds pointers, pointers to freed memory, null pointers, and pointers created with [NonNull::dangling](https://doc.rust-lang.org/nightly/std/ptr/struct.NonNull.html#method.dangling) are all dangling.

This implies that whether a pointer is dangling or not doesn't depend on the type.

These two definitions are contradictory.

**Comment by RalfJung** [comment]
"may potentially" is definitely true, it doesn't say "will always" after all.

But yeah "is dangling" uses a different definition of "dangling" than the reference. There are two reasonable definitions and we don't have a clear term to distinguish them:
- A pointer that doesn't point to allocated memory
- A pointer to type `T` that doesn't point to at least `size_of::<T>` bytes of allocated memory

**Comment by Urgau** [comment]
I just realized that `ptr::dangling` and the (recently merged) `dangling_pointers_from_temporaries` lint will both become stable in 1.84, with different uses of "dangling pointer", this is unfortunate, it would be good to clear this situation before they become stable.

**Comment by RalfJung** [comment]
`NonNull::dangling` is stable since forever, `ptr::dangling` just matches that. So I don't think there's any point in trying to pick a different name for `NonNull::dangling`.

But Cc @rust-lang/libs-api in case they disagree

**Comment by Urgau** [comment]
That's a good point. The lint however can still be changed.

**Comment by SauersML** [comment]
How about having a single "live memory for the type's size" definition (maybe in the Reference), which both the standard library docs and this lint can reference?

We could define a "dangling pointer" as "any pointer that does not point to live memory of at least `size_of::<T>()` bytes for its type," maybe noting that zero-sized types require no memory and therefore cannot be dangling in the strict sense. In `std::ptr::dangling`, we might then add a sentence explaining that for ZSTs, the pointer is not technically "dangling" (no space is required), yet we still call it "dangling" to indicate it is not intended for valid access in non-zero-size contexts.

**Comment by RalfJung** [comment]
We have a definition in the reference: https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html#dangling-pointers. So the library docs and the lint already have something they could reference, if desired.

**Comment by RalfJung** [comment]
FWIW, now that strict provenance is stable, the most accurate way to describe `ptr::dangling` is to say that it returns a pointer without provenance (like `ptr::without_provenance`).

**Comment by hkBst** [comment]
> * A pointer that doesn't point to allocated memory
> * A pointer to type `T` that doesn't point to at least `size_of::<T>` bytes of allocated memory

which of these is "pointer without provenance"?

**Comment by RalfJung** [comment]
A pointer without provenance is explained [here](https://doc.rust-lang.org/nightly/std/ptr/index.html#provenance) and [here](https://doc.rust-lang.org/nightly/std/ptr/fn.without_provenance.html).

 > A pointer to type T that doesn't point to at least `size_of::<T>` bytes of allocated memory

Note that if T is a ZST, then a pointer without provenance *does* point to at least `size_of::<T>` bytes of allocated memory.
